java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java

```diff
@@ -68,4 +70,46 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
         for (JsonObject composite : composites)
             pePlacementPreprocess.compositeColocateIdUse(composite);
     }
+
+    @SuppressWarnings("serial")
+    private void relocateHashAdders(){
+        final Set<JsonObject> hashAdders = new HashSet<>();
+        // Firstly, find each hashAdder
+        GraphUtilities.visitOnce(GraphUtilities.findStarts(graph),
+                new HashSet<BVirtualMarker>(),
+                graph,
+                (JsonObject op) -> {
+                    if (GraphUtilities.isHashAdder(op)) {
+                        hashAdders.add(op);
+                    }
+                }
+        );
+
+        for(JsonObject hashAdder : hashAdders){
+            relocateHashAdder(hashAdder);
+        }
+    }
+
+    private void relocateHashAdder(JsonObject hashAdder){
+        // Only optimize the case where $Unparallel$ is the HashAdder's only
+        // parent and the HashAdder is $Unparallel$'s only child.
+        // $Unparallel$ -> hashAdder -> $Parallel$ -> hashRemover
+        Set<JsonObject> parents = GraphUtilities.getUpstream(hashAdder, graph);
+        
+        // check if hashAdder has only one parent
+        if (parents.size() != 1) return;
+
+        JsonObject parent = parents.iterator().next();
+
+        // check if $Unparallel$ has only one child
```

Comment:
FYI - somewhat confusing here because at this point (line 104) it's not guaranteed to be  `$Unparallel$`, since you only check that at line 108. It would actually make more sense to check first it's an `$Unparallel$` before seeing how many downstream operators it has.